presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java

```diff
@@ -0,0 +1,244 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.airlift.log.Logger;
+import io.prestosql.Session;
+import io.prestosql.metadata.Split;
+import io.prestosql.metadata.TableHandle;
+import io.prestosql.operator.exchange.PageChannelSelector;
+import io.prestosql.spi.Page;
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.predicate.Domain;
+import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.spi.predicate.ValueSet;
+import io.prestosql.spi.type.Type;
+import io.prestosql.spi.type.TypeUtils;
+import io.prestosql.split.PageSourceProvider;
+import io.prestosql.sql.planner.plan.AggregationNode;
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.JoinNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.planner.plan.TableScanNode;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+public class DynamicFilter
+{
+    private static final Logger log = Logger.get(DynamicFilter.class);
+
+    private static Optional<TableScanNode> findTableScanNode(PlanNode probeNode, List<Symbol> probeSymbols)
+    {
+        while (probeNode != null) {
+            if (probeNode instanceof TableScanNode) {
+                return Optional.of((TableScanNode) probeNode);
+            }
+            if (probeNode instanceof ProjectNode) {
+                probeNode = ((ProjectNode) probeNode).getSource();
+            }
+            else if (probeNode instanceof AggregationNode) {
+                probeNode = ((AggregationNode) probeNode).getSource();
+            }
+            else if (probeNode instanceof FilterNode) {
+                probeNode = ((FilterNode) probeNode).getSource();
+            }
+            else {
+                break;
+            }
+        }
+        log.warn("createDynamicFilter: unsupported node %s", probeNode);
+        return Optional.empty();
+    }
+
+    public static Optional<DynamicFilter> create(JoinNode node, PlanNode probeNode, List<Symbol> probeSymbols, PlanNode buildNode, List<Symbol> buildSymbols, TypeProvider types)
```

Comment:
In current master `JoinNode` already contains dynamic filters `io.prestosql.sql.planner.plan.JoinNode#getDynamicFilters` and table scans contain expressions that are "dynamic filters".
Planning part is done, so I think we should reuse it.

Reply:
Good point, thanks!
Done at 39a2eb47da.